File: CoreStock/src/utils/reportFormatter.ts
import { AssessmentReport, OpportunityRow } from '../types';

const escapePipes = (value: string): string => value.replace(/\|/g, '\\|').replace(/\r?\n|\r/g, ' ');

const formatOpportunityRow = (row: OpportunityRow): string => {
    const cells = [
        row.ticker,
        row.company,
        row.sector,
        row.country,
        row.expectedDirection,
        row.timeHorizon,
        row.mechanism,
        row.investabilityScore.toString(),
        row.rationale,
        row.sources.join('; '),
    ];

    return `| ${cells.map((cell) => escapePipes(cell)).join(' | ')} |`;
};

export const formatReportAsMarkdown = (report: AssessmentReport): string => {
    const lines: string[] = [];

    lines.push('# Headline Summary');
    lines.push(report.headlineSummary.trim());
    lines.push('');

    lines.push('## Event Context');
    lines.push(report.eventContext.overview.trim());
    lines.push('');
    lines.push(`- **Timing:** ${report.eventContext.timing}`);
    lines.push(`- **Significance:** ${report.eventContext.significance}`);
    if (report.eventContext.contextPoints.length > 0) {
        lines.push('- **Key Drivers:**');
        report.eventContext.contextPoints.forEach((point) => {
            lines.push(`  - ${point}`);
        });
    }
    lines.push('');

    lines.push('## Market Impact Analysis');
    lines.push(`- **Sentiment:** ${report.marketImpact.sentiment}`);
    lines.push(`- **Macro Themes:** ${report.marketImpact.macroThemes.join('; ')}`);
    lines.push(`- **Sector Exposure:** ${report.marketImpact.sectorOutlook.join('; ')}`);
    if (report.marketImpact.horizonImpacts.length > 0) {
        lines.push('- **Time Horizons:**');
        report.marketImpact.horizonImpacts.forEach(({ horizon, outlook }) => {
            lines.push(`  - ${horizon}: ${outlook}`);
        });
    }
    lines.push('');

    lines.push('## Investment Opportunity Table');
    lines.push(
        '| Ticker | Company | Sector | Country | Expected Direction | Time Horizon | Mechanism of Impact | Investability Score | Rationale | Source(s) |',
    );
    lines.push('| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |');
    report.opportunities.forEach((row) => {
        lines.push(formatOpportunityRow(row));
    });
    lines.push('');

    lines.push('## Summary Insights');
    report.summaryInsights.forEach((insight) => {
        lines.push(`- ${insight}`);
    });
    lines.push('');

    lines.push('## Risk Note');
    lines.push(report.riskNote.trim());
    lines.push('');

    lines.push('## Citations');
    report.citations.forEach((citation) => {
        lines.push(`- ${citation}`);
    });
    lines.push('');

    lines.push(`_Generated ${new Date(report.generatedAt).toLocaleString()}_`);

    return lines.join('\n');
};

export default formatReportAsMarkdown;

File: streamlit_app/report_formatter.py
from __future__ import annotations

from datetime import datetime
from typing import Dict, List


def _escape_pipes(value: str) -> str:
    return value.replace("|", "\\|").replace("\n", " ").replace("\r", " ")


def _format_opportunity_row(row: Dict[str, object]) -> str:
    cells: List[str] = [
        str(row["ticker"]),
        str(row["company"]),
        str(row["sector"]),
        str(row["country"]),
        str(row["expected_direction"]),
        str(row["time_horizon"]),
        str(row["mechanism"]),
        str(row["investability_score"]),
        str(row["rationale"]),
        "; ".join(row.get("sources", [])),
    ]
    return "| " + " | ".join(_escape_pipes(cell) for cell in cells) + " |"


def format_report_as_markdown(report: Dict[str, object]) -> str:
    lines: List[str] = []

    lines.append("# Headline Summary")
    lines.append(str(report["headline_summary"]).strip())
    lines.append("")

    event_context = report["event_context"]
    lines.append("## Event Context")
    lines.append(str(event_context["overview"]).strip())
    lines.append("")
    lines.append(f"- **Timing:** {event_context['timing']}")
    lines.append(f"- **Significance:** {event_context['significance']}")
    context_points = event_context.get("context_points", [])
    if context_points:
        lines.append("- **Key Drivers:**")
        for point in context_points:
            lines.append(f"  - {point}")
    lines.append("")

    market_impact = report["market_impact"]
    lines.append("## Market Impact Analysis")
    lines.append(f"- **Sentiment:** {market_impact['sentiment']}")
    lines.append(f"- **Macro Themes:** {'; '.join(market_impact['macro_themes'])}")
    lines.append(f"- **Sector Exposure:** {'; '.join(market_impact['sector_outlook'])}")
    lines.append("- **Time Horizons:**")
    for item in market_impact["horizon_impacts"]:
        lines.append(f"  - {item['horizon']}: {item['outlook']}")
    lines.append("")

    lines.append("## Investment Opportunity Table")
    lines.append("| Ticker | Company | Sector | Country | Expected Direction | Time Horizon | Mechanism of Impact | Investability Score | Rationale | Source(s) |")
    lines.append("| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |")
    for row in report["opportunities"]:
        lines.append(_format_opportunity_row(row))
    lines.append("")

    lines.append("## Summary Insights")
    for insight in report["summary_insights"]:
        lines.append(f"- {insight}")
    lines.append("")

    lines.append("## Risk Note")
    lines.append(str(report["risk_note"]).strip())
    lines.append("")

    lines.append("## Citations")
    for citation in report["citations"]:
        lines.append(f"- {citation}")
    lines.append("")

    generated_at = report.get("generated_at")
    if generated_at:
        dt = datetime.fromisoformat(generated_at)
        lines.append(f"_Generated {dt.strftime('%Y-%m-%d %H:%M:%S UTC')}_")

    return "\n".join(lines)
